murdock: initial hardware-in-the-loop support

kaspar030 opened the pull request:
### Contribution description

This PR adds:

- the make target "test-murdock", which will run "make test" for the given board on murdock. (needs authentication to be set up)

- a per-application variable "TEST_WHITELIST", which is supposed to contain a list of boards on which CI should execute "make test".

- some code in ```.murdock``` that executes "make test" after successful compilation, on nightly builds or if the label "CI: run tests" is set.

Don't merge (yet), this needs a worker for every board in "TEST_WHITELIST", otherwise CI builds stall forever.

kYc0o commented on tests/unittests/Makefile:
```diff
@@ -226,12 +226,18 @@ BASELIBS += $(UNIT_TESTS:%=$(BINDIR)/%.a)
 
+info-unittests:
+	@echo $(UNIT_TESTS)
+
```
It won't be useful to give info for any test? Why only for unittests (at least in the examples you're giving).

kaspar030 replied:
Only unittests are split into tests-*. This info target outputs that information. I want to use that to split up the unittests into seperate binaries for platforms where all tests don't fit.

kaspar030 commented:
This is ready for review.

kaspar030 commented:
I'll proactively squash now that I'm happy ;)

kYc0o commented:
I added more tests. Let's see what murdock says.

aabadie commented on makefiles/murdock.inc.mk:
```diff
@@ -0,0 +1,35 @@
+
+# don't whitelist tests if there's no binary
+ifeq (1,$(RIOTNOLINK))
+  TEST_WHITELIST:=
```
Maybe call this variable TEST_BOARD_WHITELIST to be more explicit ?

kaspar030 replied:
how about TEST_ON_CI_WHITELIST?

aabadie replied:
CI_BOARD_WHITELIST ?

kaspar030 replied:
are boards in CI_BOARDS_WHITELIST compile-tested or not?

aabadie replied:
> are boards in CI_BOARDS_WHITELIST compile-tested or not?

Good point. Since you need to build before running, there's no need for it indeed. The HIL test will also check the build.

kaspar030 replied:
So TEST_ON_CI_WHITELIST is better than TEST_WHITELIST? @aabadie @cladmi?

cladmi replied:
I even missed it was a variable only for CI and not for tests in general. So yes having it explicit is good.

cladmi commented on .murdock:
```diff
@@ -91,14 +118,25 @@ compile() {
+        if [ $RUN_TESTS -eq 1 -o "$board" = "native" ]; then
+            if [ -f "${BINDIR}/.test" ]; then
+                if [ "$board" = "native" ]; then
+                    BOARD=$board make -C${appdir} test
```
Why not run the native tests also with `test-murdock` this would allow separating their output in murdock result ?

kaspar030 replied:
This is an optimization. Justs running the test saves the CI to create a job (with the attached .elf), copy job to the murdock master (over potentially slow network), then dispatch to a worker. That's an unnecessary network round trip.

The end result is that failed native tests end up as failed compile job, which is IMO cosmetic and (for now) worth the benefits of saving the network bandwidth.

cladmi commented on makefiles/murdock.inc.mk:
```diff
@@ -0,0 +1,35 @@
+#
+
+# (HACK) get actual flash binary from FFLAGS.
+FLASHFILE:=$(filter $(HEXFILE) $(ELFFILE:.elf=.bin) $(ELFFILE),$(FFLAGS))
```
It will not work for boards using `jlink.sh` as flasher as it does not get the filename from command line but from an environment variable… so there is nothing in FFLAGS.

Also having a `FLASHFILE` variable is something I am interested to fix in general.
Move the FLASHFILE outside of FFLAGS and be called as `$(FLASHER) $(FFLAGS) $(FLASHFILE)`.
We currently mix `FLASHFILE` and `HEXFILE/ELFFILE/BINFILE` names right now, and some HEXFILE are  `.bin`.

Also, for `nucleo` boards we generate only an `hex` file when a `.bin` file is needed to flash over the mounted volume. And we will need to be able to generate both outputs to handle different flashing.

kaspar030 replied:
Yeah. I started fixing that for this PR, then decided that I'd prefer to support boards working with this FLASHFILE hack first and solve the problem afterwards. Ideally, once solved elsewhere, only the "FLASHFILE" variable has to be removed,and until then, we can run tests on all boards where the hack works.

cladmi replied:
Yep, should be fixed in its proper PR.

kaspar030 commented on .murdock:
```diff
@@ -91,14 +118,25 @@ compile() {
         make -C${appdir} clean all -j${JOBS:-4}
     RES=$?
 
+    # run tests
```
Tests are only actually run on native (see below).

On other boards, the compile function runs "make test-murdock", which in turn creates a "run_test" job. thus the test jobs only created by compile().
Without massive refactoring, it has to be done like this.

kaspar030 replied:
-> wontfix

cladmi replied:
No problem to run everything at once.
I was just talking about renaming the function, which could then in bash, call two sub functions to split the function in two in the script. But its cosmetic and only used by murdock. So do as you wish with it.

kaspar030 replied:
the murdock scripts use the actual function name to sort the results. Thus if "compile" gets changed into e.g., "dispatch_compile_and_test", murdock would call that job "dispatch_compile_and_test/<application>/<board>" and also put the job's output into that folder + /output.txt. Thus I'm reluctant to change too much for now.

kaspar030 commented on makefiles/murdock.inc.mk:
```diff
@@ -0,0 +1,35 @@
+endif
+
+# create $(BINDIR)/.test file only if BOARD is in $(TEST_WHITELIST)
+link: $(BINDIR)/.test
```
> Makefile remark:

will fix

kaspar030 replied:
> The target could be set .PHONY to be re-run everytime.

I chose that option so if the Makefile changes, this removes a potentially existing .test.

kaspar030 replied:
> The target could be set .PHONY to be re-run everytime.

I chose that option so if the Makefile changes, this removes a potentially existing .test after Makefile changes.

kaspar030 commented on .murdock:
```diff
@@ -91,14 +118,25 @@ compile() {
+                    BOARD=$board make -C${appdir} test
+                    RES=$?
+                else
+                    if is_in_list "$board" "$TEST_BOARDS_AVAILABLE"; then
```
yup

kaspar030 replied:
done

kaspar030 commented:
> I added more tests. Let's see what murdock says.

@kYc0o please open another PR for extending the enabled tests, so we split infrastructure and its application. Otherwise, this PR is unnecessarily blocked until all tests complete successfully.

kYc0o commented:
> @kYc0o please open another PR for extending the enabled tests, so we split infrastructure and its application. Otherwise, this PR is unnecessarily blocked until all tests complete successfully.

OK, I pushed here since it was already adding some tests, and I wasn't sure it would work also for a separated PR. I'll open another one then.

kaspar030 commented:
- changed "TEST_WHITELIST" to "TEST_ON_CI_WHITELIST".

IMO whe should get this in. @cladmi care to ACK?

kaspar030 commented:
@smlng @cgundogan maybe you could take a look?

kaspar030 commented:
@cladmi *ping* ACK now and I'll bring a random berlin bottled craft beer to the next pizza day!

cladmi commented:
Its good for me, I let you merge when you want if you want more feedback.

kaspar030 commented:
> Its good for me, I let you merge when you want if you want more feedback.

Let's see how the nightly looks like.

kaspar030 commented:
Ahja, and let's not tell everyone that on-hardware test can be triggered using the "CI: run tests" label, and test this a bit with selected PR's first. :smirk:

kaspar030 commented:
> (which beer ? :p)

Surprise!
